File: Sources/RZFlight/Waypoint.swift
//
//  Waypoint.swift
//  RZFlight
//
//  Named navigation waypoint (5-letter codes and NAVAIDs like VOR/DME/NDB).
//

import Foundation
import CoreLocation
import FMDB

/// A named navigation waypoint with coordinates.
///
/// Waypoints come from sources like Eurocontrol's FRA (Free Route Airspace)
/// points list. Used alongside airports for route resolution.
public struct Waypoint: Codable, Sendable {

    /// Waypoint name (e.g., "BILGO", "REM") — primary key
    public let name: String

    /// Pre-computed lowercase name for fast search
    public let nameLower: String

    /// Latitude in decimal degrees
    public let latitude: Double

    /// Longitude in decimal degrees
    public let longitude: Double

    /// Point type: "5LNC", "VOR", "DME", "VORDME", "NDB", "VORTAC", "NDBDME", "LOCATOR"
    public let pointType: String?

    /// Comma-separated FIR ICAO codes
    public let firCodes: String?

    /// Level availability (e.g., "FL195 / FL660")
    public let levelAvailability: String?

    /// Data source identifier
    public let source: String?

    /// Coordinates as CLLocationCoordinate2D
    public var coordinate: CLLocationCoordinate2D {
        return CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
    }

    /// Whether this is a NAVAID (VOR, DME, NDB, etc.) rather than a 5-letter code
    public var isNavaid: Bool {
        guard let pt = pointType else { return false }
        return pt != "5LNC"
    }

    /// FIR codes as an array
    public var firList: [String] {
        guard let codes = firCodes else { return [] }
        return codes.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces) }
    }

    // MARK: - Initialization

    public init(name: String, latitude: Double, longitude: Double,
                pointType: String? = nil, firCodes: String? = nil,
                levelAvailability: String? = nil, source: String? = nil) {
        self.name = name
        self.nameLower = name.lowercased()
        self.latitude = latitude
        self.longitude = longitude
        self.pointType = pointType
        self.firCodes = firCodes
        self.levelAvailability = levelAvailability
        self.source = source
    }

    /// Initialize from a database result set.
    public init?(res: FMResultSet) {
        guard let name = res.string(forColumn: "name") else { return nil }

        self.name = name
        self.nameLower = name.lowercased()
        self.latitude = res.double(forColumn: "latitude_deg")
        self.longitude = res.double(forColumn: "longitude_deg")
        self.pointType = res.string(forColumn: "point_type")
        self.firCodes = res.string(forColumn: "fir_codes")
        self.levelAvailability = res.string(forColumn: "level_availability")
        self.source = res.string(forColumn: "source")
    }

    // MARK: - Search

    public func contains(_ searchText: String) -> Bool {
        return nameLower.contains(searchText.lowercased())
    }

    func matches(_ needle: String) -> Bool {
        return nameLower.contains(needle.lowercased())
    }

    // MARK: - Codable

    enum CodingKeys: String, CodingKey {
        case name
        case latitude = "latitude_deg"
        case longitude = "longitude_deg"
        case pointType = "point_type"
        case firCodes = "fir_codes"
        case levelAvailability = "level_availability"
        case source
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        let name = try container.decode(String.self, forKey: .name)
        self.name = name
        self.nameLower = name.lowercased()
        self.latitude = try container.decode(Double.self, forKey: .latitude)
        self.longitude = try container.decode(Double.self, forKey: .longitude)
        self.pointType = try container.decodeIfPresent(String.self, forKey: .pointType)
        self.firCodes = try container.decodeIfPresent(String.self, forKey: .firCodes)
        self.levelAvailability = try container.decodeIfPresent(String.self, forKey: .levelAvailability)
        self.source = try container.decodeIfPresent(String.self, forKey: .source)
    }
}

// MARK: - Protocol Conformances

extension Waypoint: Hashable {
    public static func == (lhs: Waypoint, rhs: Waypoint) -> Bool {
        return lhs.name == rhs.name
    }

    public func hash(into hasher: inout Hasher) {
        hasher.combine(name)
    }
}

extension Waypoint: Identifiable {
    public var id: String { name }
}

extension Waypoint: CustomStringConvertible {
    public var description: String {
        if let pt = pointType, pt != "5LNC" {
            return "\(name) (\(pt))"
        }
        return name
    }
}

File: euro_aip/euro_aip/models/waypoint.py
"""
Waypoint model for named navigation points (5-letter codes and NAVAIDs).

Waypoints are distinct from airports — they represent named fixes, VORs, DMEs,
NDBs, and other navigation points used in route strings.
"""

from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional, Dict, Any

from euro_aip.models.navpoint import NavPoint


@dataclass
class Waypoint:
    """A named navigation waypoint with coordinates.

    Waypoints come from sources like Eurocontrol's FRA (Free Route Airspace)
    points list. They have a `.navpoint` property for distance calculations,
    just like Airport does.
    """

    name: str  # "BILGO", "REM", "ABADI"
    latitude_deg: float
    longitude_deg: float
    point_type: Optional[str] = None  # "5LNC", "VOR", "DME", "VORDME", "NDB", "VORTAC", "NDBDME", "LOCATOR"
    fir_codes: Optional[str] = None  # Comma-separated FIR ICAOs e.g. "LFFF,LFBB"
    level_availability: Optional[str] = None  # e.g. "FL195 / FL660"
    source: str = "eurocontrol_fra"
    source_id: str = ""  # Unique per candidate, e.g. "fra:LFFF", "opennav:UK", "ourairports:GB"
    created_at: datetime = field(default_factory=datetime.now)
    updated_at: datetime = field(default_factory=datetime.now)

    _navpoint: Optional[NavPoint] = field(default=None, init=False, repr=False, compare=False)

    @property
    def navpoint(self) -> NavPoint:
        """Get NavPoint representation for distance calculations."""
        if self._navpoint is None:
            self._navpoint = NavPoint(
                latitude=self.latitude_deg,
                longitude=self.longitude_deg,
                name=self.name,
            )
        return self._navpoint

    @property
    def is_navaid(self) -> bool:
        """True if this waypoint is a NAVAID (VOR, DME, NDB, etc.) rather than a 5LNC."""
        return self.point_type is not None and self.point_type != "5LNC"

    @property
    def fir_list(self) -> list:
        """Get FIR codes as a list."""
        if not self.fir_codes:
            return []
        return [f.strip() for f in self.fir_codes.split(",") if f.strip()]

    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary for serialization."""
        return {
            "name": self.name,
            "latitude_deg": self.latitude_deg,
            "longitude_deg": self.longitude_deg,
            "point_type": self.point_type,
            "fir_codes": self.fir_codes,
            "level_availability": self.level_availability,
            "source": self.source,
            "source_id": self.source_id,
            "created_at": self.created_at.isoformat() if self.created_at else None,
            "updated_at": self.updated_at.isoformat() if self.updated_at else None,
        }

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "Waypoint":
        """Create from dictionary."""
        created_at = data.get("created_at")
        if isinstance(created_at, str):
            created_at = datetime.fromisoformat(created_at)
        updated_at = data.get("updated_at")
        if isinstance(updated_at, str):
            updated_at = datetime.fromisoformat(updated_at)

        return cls(
            name=data["name"],
            latitude_deg=data["latitude_deg"],
            longitude_deg=data["longitude_deg"],
            point_type=data.get("point_type"),
            fir_codes=data.get("fir_codes"),
            level_availability=data.get("level_availability"),
            source=data.get("source", "eurocontrol_fra"),
            source_id=data.get("source_id", ""),
            created_at=created_at or datetime.now(),
            updated_at=updated_at or datetime.now(),
        )

    def __repr__(self):
        type_str = f" ({self.point_type})" if self.point_type else ""
        return f"Waypoint({self.name}{type_str}, {self.latitude_deg:.4f}, {self.longitude_deg:.4f})"
